src/rpc/mining.cpp

```diff
@@ -58,63 +58,144 @@ unsigned int ParseConfirmTarget(const UniValue& value)
+                pbFirst = pb;
+            }
+        }
+        // Check if we should continue
+        if (((0 != type.compare("sha256d")) || (0 != type.compare("randomx")) || (0 != type.compare("progpow"))) &&
+            (pb->GetBlockTime() < Params().PowUpdateTimestamp()))
+        {
+            // Don't keep searching, abort now
+            break;
+        }
+    }
+
+    // only one was found
+    if (!pbFirst || !pbLast || (pbFirst == pbLast)) return 0;
+
+    int64_t minTime = pbFirst->GetBlockTime();
+    int64_t maxTime = pbLast->GetBlockTime();
+
     // In case there's a situation where minTime == maxTime, we don't want a divide by zero exception.
     if (minTime == maxTime)
         return 0;
 
-    arith_uint256 workDiff = pb->nChainWork - pb0->nChainWork;
+    arith_uint256 workDiff = pbLast->nChainWork - pbFirst->nChainWork;
     int64_t timeDiff = maxTime - minTime;
 
     return workDiff.getdouble() / timeDiff;
 }
 
 static UniValue getnetworkhashps(const JSONRPCRequest& request)
 {
-    if (request.fHelp || request.params.size() > 2)
+    if (request.fHelp || request.params.size() > 3)
         throw std::runtime_error(
-            "getnetworkhashps ( nblocks height )\n"
+            "getnetworkhashps ( nblocks height algo )\n"
             "\nReturns the estimated network hashes per second based on the last n blocks.\n"
             "Pass in [blocks] to override # of blocks, -1 specifies since last difficulty change.\n"
-            "Pass in [height] to estimate the network speed at the time when a certain block was found.\n"
+            "Pass in [height] to estimate the network hash at the time when a certain block was found.\n"
+            "Pass in [algo] to estimate the network hash for a different algo then your wallet is set to.\n"
             "\nArguments:\n"
             "1. nblocks     (numeric, optional, default=120) The number of blocks, or -1 for blocks since last difficulty change.\n"
             "2. height      (numeric, optional, default=-1) To estimate at the time of the given height.\n"
+            "3. algo        (numeric, optional, default="+GetMiningType(GetMiningAlgorithm(), false, false)+") Algo to calculate [randomx, sha256d, progpow, xr16t, PoS].\n"
             "\nResult:\n"
             "x             (numeric) Hashes per second estimated\n"
             "\nExamples:\n"
             + HelpExampleCli("getnetworkhashps", "")
             + HelpExampleRpc("getnetworkhashps", "")
        );
 
-    LOCK(cs_main);
-    return GetNetworkHashPS(!request.params[0].isNull() ? request.params[0].get_int() : 120, !request.params[1].isNull() ? request.params[1].get_int() : -1);
+
+    if (!(request.params[2].isNull())) {
+        std::string algo = request.params[2].get_str();
+        // Check if it's a mining algorithm
+        if (!SetMiningAlgorithm(algo, false))
+            // check that it's not one of the other two
+            if (algo.compare("PoS") && algo.compare("xr16t"))
```

Comment:
Should this be an or?  ||

Reply:
compare() is confusing; it returns '0' if it matches.  so in C++, `if compare()` means if it doesn't match.  So here, we're checking that it "doesn't match PoS and doesn't match xr16t".  So rolling the whole flow together:

If SetMiningAlgorithm failed, it's either a typo, or it's PoS or it's xr16t.  So if it's not a mining algo, and it's not PoS and it's not xr16t, then we'll throw an error.